docs/rfcs/uuidgen-refactor.md

```diff
@@ -0,0 +1,222 @@
+
+   // or for `no features`
+
+   struct MyTime {};
+   impl TimeSource for MyTime {
+   	fn next_unix_time() -> (u64, u32) {
+	  ...
+	}
+   }
+   struct MySequentialGen {};
+   impl SequenceSource for MySequentialGen {
+        fn next_u64() -> u64 {
+	  ...
+	}
+   }
+   let bldr: UuidV1GenBuilderNoClockNoSeq = UuidGenBuilder::new_v1();
+   let bldr: UuidV1GenBuilderNoClock<MyTime> = bldr.with_seq_source(MySequentialGen {});
+   let gen: UuidV1Gen<MyTime, MySequentialGen> = bldr.with_time_source(MyTimeSource {}).finish();
+   let u2 = gen.next();
+
+   // Note that all of these types are for illustration purposes really a user would just do
+
+   let infinite_ids = UuidV1GenBuilder::new_v1()
+       .with_seq_source(MySequentialGen {})
+       .with_time_source(MyTimeSource {})
+       .finish()
+
+   let hundred_ids = infinite_ids.take(100).collect::<Vec<UuidV1>>();
+```
+
+So there would be probably 4 typestate structs per uuid version, and 8 versions, but they'd be very simple structs, as they'd exist to ensure that only the correct setters could be used.
+
+The methods each struct impls would vary based on the requirements of its Uuid version, and what built-in functionality the user has enabled via features.
+
+e.g.
+
+```rust
+struct UuidV6GenBuilderNoSeq<TS: TimeSource> {
+  clock_source: TS
+}
+
+impl<TS: TimeSource> UuidV6GenBuilderNoSeq<TS> {
+  #[cfg(feature = "thread-rng"]
+  pub fn with_thread_local_rng(self) -> UuidV6GenBuilder {
+  	let rng = rand::thread_rng();
+	UuidV6GenBuilder {
+	  seq_source: rng,
+	  clock_source: self.clock_source,
+	}
+  }
+
+  pub fn with_seq_source<S: SequenceSource>(self, src: S) -> UuidV6GenBuilder {
+       UuidV6GenBuilder {
+          seq_source: src,
+	  clock_source: self.clock_source,
+       }
+  }
+}
+```
+
+
+### New Supporting Traits:  TimeSource and SequenceSource
+
+TimeSource and SequenceSource would be a disaggregation of the current ClockSequence trait
+
+SequenceSource would be intended to serve both Monotonically Increasing Counters as well as Random Number Generators.
+Both are (probably ) stateful sequence generators.
+
+I am undecided how best to represent the outputs of the trait's methods.
+There are 2 approaches:
+
+1. Associated Type: Output
+One approach would be to just have an associated type, Output.  This would certainly be generic enough to handle all cases, but it would be a bit more hassle to implement for both maintainers and users.
+Though we could have on-hand implementations of the most common types and approaches.
+
+```rust
+trait TimeSource {
+    type Output;
+    fn next(&mut self) -> Self::Output;
+}
+
+// Note that SequenceSource is parameterized over TimeSource::Output because some implementations might want to track the previous Clock output to see if it was a duplicate before increasing their counter.
+trait SequenceSource<T: TimeSource> {
+    type Output;
+    fn next(&mut self, tm: <T as TimeSource>::Output) -> Self::Output;
```

Comment:
Rather than making this part of `SequenceSource`, what about requiring that `TimeSource::Output` implements `Eq`, and then deciding inside the UUID library whether to check for duplicates or not?

I'm also wondering to what extent it makes sense for `TimeSource::Output` to be parameterized, without having some notion of precision with which to turn it into the component needed for a UUID? Or is the expectation that the library will nonetheless rely on specific output types rather than being generic?

Reply:
I definitely like the idea of removing any awareness of `TimeSource` from `SequenceSource`.   However, whether or not the `SequenceSource` checks the time for duplicates isn't really a version-specific thing, it's a preference of the implementation.  If we could pass in some notion of "there was a conflict" into `SequenceSource::next()` that would be fine by me.  Maybe an `Option<bool>`

Regarding Output precision:   My assumption was that the precision required would be version-specific, so the `Output` would always be parameterized by each `UuidV#` Implementation.   The only exception to this, though, would be UuidV8, which violates this rule.   

TBH, this GenBuilder API was inspired by the idea that one could use a builder to make a highly tuned UuidV8 implementation by choosing how many bits they wanted for seconds, frac-seconds, random and sequence.  However, there might be better/safer ways to accommodate that level of flexibility in the API design.  (I was going to add the precision/size check at runtime in the Builder construction)

Reply:
@rrichardson That does seem like perhaps more flexibility than necessary.

And yeah, passing in "there was a conflict" seems like an improvement, at least.